tests/test_url.py

```diff
@@ -799,13 +799,25 @@ def test_div_with_dots():
         pytest.param(
             "/path/", ("to",), "http://example.com/path/to", id="path-with-slash"
         ),
+        pytest.param(
+            "/path", ("",), "http://example.com/path/", id="path-add-trailing-slash"
+        ),
         pytest.param(
             "/path?a=1#frag",
             ("to",),
             "http://example.com/path/to",
             id="cleanup-query-and-fragment",
         ),
         pytest.param("", ("path/",), "http://example.com/path/", id="trailing-slash"),
+        pytest.param(
+            "",
+            (
+                "path",
+                "",
+            ),
+            "http://example.com/path/",
+            id="trailing-slash-empty-string",
+        ),
```

Comment:
Normalized path do not have double slashes?

And from a technical point of view, 
`("path/", "")` is the same as `("path", "", "")`
So the third would be bad as well.

The requested behavior does not match the `test_joinpath_relative[empty-element]` - the test you broke.

https://github.com/aio-libs/yarl/blob/f9e3b1b477cd3a41746c89be2a0cb334f7b87183/tests/test_url.py#L835

Reply:
Are you sure these would be a normalised paths? 
Even if so, the lib should allow to generate non-normalised paths, since many web servers allow to differentiate for paths with and without slashes at the end for example.

Reply:
according to RFC

Reply:
According to [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986.html#section-3.3), the path is allowed to have empty strings as segment of the path, as defined in the grammar.

Reply:
You are right wrt. to empty segments are allowed, additionally RFC 3986 normalization does not include normalizing empty segments.

Still … technically those would have to be equivalent
```
# op == segments == str(…)
"path" == ("path",) == "path"
"path/" == ("path", "") == "path/"
"path" / "a" == ("path", "a") == "path/a"
"path/" / "a" == ("path", "", "a") == "path//a"
```
and this requires testing for a trailing / before joining to avoid "path//a"

therefore I'd normalize empty-segments as well.
```
"path/" / "a" == ("path", "", "a") == "path/a"
```

Reply:
Yeah, the spec doesn't seem to consider such paths to be equivalent. The only detail that matters is that an empty path should be normalised to `/`, which yarl doesn't seem to do right now (so `yarl("http://foo")` should produce `http://foo/`).

I'd argue that the result of `"path/" / "a"` should be `path/a` though (and is not really related to the spec). Certainly if we're requiring a `/ ""` to include the trailing slash, then it makes sense for each intermediate part to be stripped. But, even if it's not required, I think it's surprising for the user if the operator results in a double slash (use a string literal if you actually want a double slash).

Reply:
In other words, the `/` operator does not represent a literal slash, but is a method to join two parts of a URL together. So, if there's already a slash there, it shouldn't add another to complete the join.

Reply:
My point is that YARL should allow building URLs that are not “normalised”, as many use cases require that.
Also keep in mind that URLs path behaviour should not match OS paths, as they are for a different purpose. Mostly all web servers will consider `example.com/foo/../bar` to be very different from `example.com/bar`, as the application logic does not necessarily relate to an OS path logic.

Reply:
> In other words, the `/` operator does not represent a literal slash, but is a method to join two parts of a URL together. So, if there's already a slash there, it shouldn't add another to complete the join.

[5.2.3](https://datatracker.ietf.org/doc/html/rfc3986.html#section-5.2.3).  Merge Paths

Basically - a join has to remove the last segment:
```python
>>> urllib.parse.urljoin("http://host/path/1", "a")
# 'http://host/path/a'
```

```python
>>> URL("http://host/path/1") / "a"
# URL('http://host/path/1/a')
```

Reply:
The / operator should work more closely to al literal `/` than not, otherwise if you follow that spec 5.2.3 you would have that 
`URL('/')/“foo”/“bar”/“baz” == URL('/baz')` which is definitely not what you want.

Reply:
> Mostly all web servers will consider `example.com/foo/../bar` to be very different from `example.com/bar`

On this point, the spec seems pretty clear. These _are_ the same path and they should be transformed before making the request. Whether we process it immediately in yarl, or before making the request in aiohttp, the request will always be for 'example.com/bar'. We should never send the original with dots. Note that this part isn't only under the normalisation section, which is optional, it appears to be a mandatory behaviour under the resolution section (i.e. if the server behaves differently, it should only be because they don't know how to parse the dots as they should never be receiving them in the first place. You can see this behaviour in browsers too, put that link on a web page and the browser shows the transformed URL before you even click it).

So, if we want to proceed, I think the design of yarl is to normalise as much as possible according to the specs.
For this PR, I think we should be settling on:

- Extra slashes in a string should _not_ be removed/normalised (the spec does suggest/allow this).
- The `/` operator should not introduce a double slash, which would be unexpected behaviour, but should add one when none is present in the strings.

In the future, with another PR, I think the only normalisation step currently missing is https://datatracker.ietf.org/doc/html/rfc3986.html#section-6.2.3
Which would include (for HTTP/HTTPS schemes) setting the path to `/` when empty and removing the port number when matching the default 80/443.

Reply:
Does this match your intention?

- "path" / "a" -> "path/a"
- "path" / "./a" -> "path/a"
- "path/" / "a" -> "path/a"
- "path/ " / "./a" -> "path/a"
- "path" / ".//a" -> "path//a"
- "path/" / ".//a" -> "path//a"
- "path//" / "a" -> "path//a"
- "path//" / "./a" -> "path//a"
- "path//" / ".//a" -> "path////a"

Reply:
@youtux how to you think about:
https://github.com/commonism/yarl/commit/56db9d3ce2f926da28e9e89fe68b39d8f856aef8

it passes all current tests and satisfies the requirements set.

Reply:
I think that all looks correct to me, with no opinion on whether the last example should have 3 or 4 slashes.

Reply:
Last example is meant to have 3 slashes - 4 would be a special case.
I've refactored things a little to match the wording in rfc3986 and added more tests to document expected behavior.
URL.join is inconsistent to URL.joinpath now, it uses urlib.parse.join, stripping empty segments.

Reply:
adding a consistent URL.join & rfc3986 port normalization
https://github.com/commonism/yarl/compare/youtux-fix-join-trailing-slash...commonism:yarl:rfc3986-port-normalization?expand=1

How do you want this to proceed @youtux?

Reply:
I think you should open a new pull request since it's more complete.
Also, I'd split other changes you made in different PRs (like the query params changes)